Common/Orders/Fills/FillModel.cs

```diff
@@ -114,8 +114,15 @@ public virtual OrderEvent MarketFill(Security asset, MarketOrder order)
             // make sure the exchange is open/normal market hours before filling
             if (!IsExchangeOpen(asset, false)) return fill;
 
-            //Order [fill]price for a market order model is the current security price
-            fill.FillPrice = GetPricesCheckingPythonWrapper(asset, order.Direction).Current;
+            var prices = GetPricesCheckingPythonWrapper(asset, order.Direction);
+            var pricesEndTime = prices.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+
+            // do not fill on stale data
+            if (pricesEndTime < order.Time) return fill;
+
+            //Order [fill]price for a market order model is the current security price or the open price of the next bar
+            fill.FillPrice = pricesEndTime > order.Time ? prices.Open : prices.Current;
+
```

Comment:
Very interesting insights. Regarding live mode -- completely agreed should fill using the tick stream, the data for which should already be in the `SecurityCache`. Also, since live-mode pulses the algo loop _at least_ once a second, we're guaranteed to have the fill methods invoked within a second independent of the configured resolution.

One concern I have w/ pulsing open prices at the start of every bar is it might have the effect of cutting performance almost in half in backtest since we'll be spinning the algo manager loop 2x (once for open tick and once for bar). Backtesting sounds like the tougher case here, for live, was thinking could add an event that is raised each time the security's real-time price is updated and then fill models could add an event handler, wait for the next tick, create the fill, and then remove their event handler. I suppose we could _mock_ real-time price in backtest by doing something like linear regression between open/close of the bars? This would keep the 'look-ahead' logic in the data feed and prevent it from needing to creep into algo manager and other parts of the system. Certainly not  a solid idea, but just trying to add some off-the-wall ideas into the mix :)

Also, getting algo manager to look ahead might be tough, although I suppose he could always have the _next_ time step in hand and the 'current' loop would always be processing the _previous_ time step, so he'd be holding on to 2 time steps each loop (current/next).

Happy to bounce ideas around if you guys would like to chat :)